### PR TITLE
Tests passing, issue #50 finished by the way

### PR DESCRIPTION
**Pull request by cymerrad**

### CODE REVIEW

File: jest.config.js [hunk 1]
```diff
@@ -4,6 +4,7 @@ module.exports = {
   // Note: '*.test.ts' intentionally omitted; '*.spec.ts' stands out in VSCode as it is yellow
   // Also, no tsx files.
   testRegex: "(/__tests__/.*|(\\.|/)(spec))\\.ts$",
+  // testRegex: "(/__tests__/(parsers|resolvers)/.*\\.spec)\\.ts$",
```
Comment (ritave) [inline jest.config.js]: Remove if not used, if somebody wants it back - they can grep the git history

File: src/resolvers/backtrackfsresolver.ts [hunk 2]
```diff
@@ -1,22 +1,26 @@
...
     }
 
-    const cwd = ctx.cwd || process.cwd();
+    const cwd = ctx ? ctx.cwd || process.cwd() : process.cwd();
 
     let previous: string = path.resolve(cwd, "./");
     let current: string = previous;
     do {
-      const result = await fsResolver(path.join(current, pathPrefix, resolvePath), ctx);
+      const shotInTheDark = path.join(current, pathPrefix, resolvePath);
```
Comment (ritave) [inline src/resolvers/backtrackfsresolver.ts]: Naming

Comment (ritave) [reply src/resolvers/backtrackfsresolver.ts]: Or remove debug function altogether

File: src/resolvers/fsresolver.ts [hunk 3]
```diff
@@ -1,15 +1,23 @@
 import * as fs from "fs";
 import * as path from "path";
-import { promisify } from "util";
 import { SubResolver } from ".";
 import { ResolverContext } from "./subresolver";
 
-const statAsync = promisify(fs.stat);
+const statAsync = (path: string): Promise<fs.Stats> =>
```
Comment (ritave) [inline src/resolvers/fsresolver.ts]: Fs resolver is used only in node context - same as stat, promisify can stay.

File: src/resolvers/githubresolver.ts [hunk 4]
```diff
@@ -1,30 +1,47 @@
...
 
-// hosted-git-info is a godsend, but it doesn't support specific files
-// use a hack to capture third slash a file skipping commitish
-// 1st group - protocol, location, owner, repo
-// 2nd group - file inside repo
-// 3rd group - comittish
-const FILE_LOCATION_REGEX = /^((?:.+:\/\/)?[^:/]+[/:][^/]+[/][^/]+)[/](.+?)(#.+)?$/;
+const BROWSER_LINK_REGEX = /^https?:\/\/github\.com\/([^/]+)\/([^/]+)\/blob\/((?:[^/]+[/])*[^/]+)$/;
+const GITHUB_URI_REGEX = /^github:([^/]+)\/([^/]+)\/((?:[^/#]+[/])*[^/#]+?)(#\w+)?$/;
+const WEIRD_BROWSER_LINK_REGEX = /^https?:\/\/github\.com\/([^/]+)\/([^/]+)\/((?:[^/]+[/])*[^/]+)$/; // remix
```
Comment (ritave) [inline src/resolvers/githubresolver.ts]: Naming and explanation

File: src/resolvers/githubresolver.ts [hunk 5]
```diff
@@ -1,30 +1,47 @@
...
 
 const debug = Debug("resolverengine:githubresolver");
 
-// hosted-git-info is a godsend, but it doesn't support specific files
-// use a hack to capture third slash a file skipping commitish
-// 1st group - protocol, location, owner, repo
-// 2nd group - file inside repo
-// 3rd group - comittish
-const FILE_LOCATION_REGEX = /^((?:.+:\/\/)?[^:/]+[/:][^/]+[/][^/]+)[/](.+?)(#.+)?$/;
+const BROWSER_LINK_REGEX = /^https?:\/\/github\.com\/([^/]+)\/([^/]+)\/blob\/((?:[^/]+[/])*[^/]+)$/;
```
Comment (ritave) [inline src/resolvers/githubresolver.ts]: Those regexes are hard to understand, include an explanation for each of those

Comment (ritave) [reply src/resolvers/githubresolver.ts]: Also they remove support for comittish (`github:user/repo#branch`), gitlab and bitbucket

File: src/resolvers/githubresolver.ts [hunk 6]
```diff
@@ -1,30 +1,47 @@
...
+const GITHUB_URI_REGEX = /^github:([^/]+)\/([^/]+)\/((?:[^/#]+[/])*[^/#]+?)(#\w+)?$/;
+const WEIRD_BROWSER_LINK_REGEX = /^https?:\/\/github\.com\/([^/]+)\/([^/]+)\/((?:[^/]+[/])*[^/]+)$/; // remix
 
 // TODO(ritave): Support private repositories
 export function GithubResolver(): SubResolver {
   return async (what: string): Promise<string | null> => {
-    const fileMatch = what.match(FILE_LOCATION_REGEX);
-    if (!fileMatch) {
-      return null;
+    // GitInfo returned invalid results (ask me about this: %2F)
```
Comment (ritave) [inline src/resolvers/githubresolver.ts]: Talked about this - remove comment

File: src/resolvers/githubresolver.ts (hunk 5 above)
Comment (ritave) [reply src/resolvers/githubresolver.ts]: Added into the test excel

File: src/resolvers/githubresolver.ts [hunk 7]
```diff
@@ -1,30 +1,47 @@
...
+
+    const fileMatchLink = what.match(BROWSER_LINK_REGEX);
+    if (fileMatchLink) {
+      const [, owner, repo, commitAndFile] = fileMatchLink;
+      const gitRawUrl = `https://raw.githubusercontent.com/${owner}/${repo}/${commitAndFile}`;
```
Comment (ritave) [inline src/resolvers/githubresolver.ts]: Duplication of work - hosted-git-info resolves into raw urls - https://github.com/npm/hosted-git-info#methods

> For example `hostedGitInfo.fromUrl("git@github.com:npm/hosted-git-info.git#v1.0.0").file("package.json")` would return `https://raw.githubusercontet.com/npm/hosted-git-info/v1.0.0/package.json`

File: src/resolvers/githubresolver.ts [hunk 8]
```diff
@@ -1,30 +1,47 @@
...
+      debug("Resolved uri to:", gitRawUrl);
+      return gitRawUrl;
     }
-    const [, url, file, comittish] = fileMatch;
-    const gitInfo = GitInfo.fromUrl(url + (comittish || ""));
-    if (!gitInfo) {
-      return null;
+
+    const fileMatchWeird = what.match(WEIRD_BROWSER_LINK_REGEX);
```
Comment (ritave) [inline src/resolvers/githubresolver.ts]: naming

File: src/resolvers/githubresolver.ts [hunk 9]
```diff
@@ -1,30 +1,47 @@
...
+        commit = "master";
+      } else {
+        commit = hashCommit.substr(1);
+      }
+      const gitRawUrl = `https://raw.githubusercontent.com/${owner}/${repo}/${commit}/${file}`;
+      debug("Resolved uri to:", gitRawUrl);
+      return gitRawUrl;
+    }
+
+    return null;
```
Comment (ritave) [inline src/resolvers/githubresolver.ts]: Generally this mostly does the same thing as insides of hosted-git-info, please read the sources of it and try to use it as much as possible

File: src/resolvers/ipfsresolver.ts [hunk 10]
```diff
@@ -0,0 +1,8 @@
+import { SubResolver } from ".";
```
Comment (ritave) [inline src/resolvers/ipfsresolver.ts]: Don't add things that don't work into `master`

File: src/resolvers/noderesolver.ts [hunk 11]
```diff
@@ -1,24 +1,10 @@
...
-        requirePaths.push(currentPath);
-        lastPath = currentPath;
-        currentPath = path.dirname(currentPath);
-      } while (currentPath !== lastPath);
-      return require.resolve(what, { paths: requirePaths });
-    } catch (e) {
-      debug("Node's require.resolve failed, returning null", e);
-      return null;
-    }
+  const backtrack = BacktrackFsResolver("node_modules");
```
Comment (ritave) [inline src/resolvers/noderesolver.ts]: An edge case might happen if the package is in global node_modules

Comment (ritave) [reply src/resolvers/noderesolver.ts]: Added to test excel

File: src/resolvers/swarmresolver.ts [hunk 12]
```diff
@@ -0,0 +1,8 @@
+import { SubResolver } from ".";
```
Comment (ritave) [inline src/resolvers/swarmresolver.ts]: Don't add things into master that don't work

File: src/solidity/ethpmresolver.ts [hunk 13]
```diff
@@ -1,28 +1,41 @@
...
+const debug = Debug("resolverengine:ethpmresolver");
 
 // 1st group - package name
 // 2nd group - contract path
-const FILE_LOCATION_REGEX = /^(.+?)\/(.+)$/;
+const FILE_LOCATION_REGEX = /^([^/]+)\/(.+)$/;
 
-export function EthPmResolver(epmDirectory: string = "installed_contracts/"): SubResolver {
-  const fsResolver = BacktrackFsResolver(epmDirectory);
+const prefixT = "installed_contracts";
```
Comment (ritave) [inline src/solidity/ethpmresolver.ts]: Naming

File: src/solidity/ethpmresolver.ts [hunk 14]
```diff
@@ -1,28 +1,41 @@
...
+
+  return async (what: string, ctx?: ResolverContext): Promise<string | null> => {
     const fileMatch = what.match(FILE_LOCATION_REGEX);
     if (!fileMatch) {
       return null;
     }
 
-    // In case it actually spells out the "contracts/" folder
-    let result = await fsResolver(what, ctx);
+    // const [, packageName, internalPath] = fileMatch;
```
Comment (ritave) [inline src/solidity/ethpmresolver.ts]: Unused comment

File: src/solidity/ethpmresolver.ts [hunk 15]
```diff
@@ -1,28 +1,41 @@
...
+
+    // let result = backtrackT(path.join(packageName, prefixT + "/", internalPath), ctx);
```
Comment (ritave) [inline src/solidity/ethpmresolver.ts]: Unused comment

File: test/MockedFs.ts [hunk 16]
```diff
@@ -0,0 +1,10 @@
+const mockedFS = {
+  "/path/to/file.txt": "contents of /path/to/file.txt", 
+  "/dir/node_modules/package/some/file.txt": "contents of package/some/file.txt", 
+  "/dir/project/sumdir/main.lol": "DA WAE", 
+  "/eth/contracts/zeppelin/contract.sol": "solidity stuff", 
+  "/eth/installed_contracts/ledzeppelin/othercontract.sol": "lol the sol", 
+  "/eth/someproject/file.lol": "E"
+}
+
+export default mockedFS;
```
Comment (ritave) [inline test/MockedFs.ts]: newline

File: test/__tests__/parsers/UrlParser.spec.ts [hunk 17]
```diff
@@ -0,0 +1,35 @@
+import nock from "nock";
+import { UrlParser } from "../../../src/parsers/urlparser";
+
+// input, output
+const incorrectUrls = [["", null], ["www.google.com", null]];
+const correctUrls = [["http://example.com/some/file.txt", "http://example.com/some/file.txt"]];
+const data = incorrectUrls.concat(correctUrls);
+
+describe(UrlParser, () => {
+  const parser = UrlParser(); // unintuitive af
```
Comment (ritave) [inline test/__tests__/parsers/UrlParser.spec.ts]: This comment is unnecessary

File: test/__tests__/resolvers/EthPmResolver.spec.ts [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+  });
+  afterEach(() => {
+    vol.reset();
+  });
+
+  it.each(data)("testing %o in context %o", async (input, context, output) => {
+    const actualOutput = await subject(input, context);
+    expect(actualOutput).toBe(output);
+
+    // IMPLEMENT ME
```
Comment (ritave) [inline test/__tests__/resolvers/EthPmResolver.spec.ts]: Make the tests throw an error and skip those tests so that they show as different result than green - https://stackoverflow.com/a/51561357

File: test/__tests__/resolvers/FsResolver.spec.ts [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+  });
+  afterEach(() => {
+    vol.reset();
+  });
+
+  it.each(data)("testing %o in context %o", async (input, context, output) => {
+    const actualOutput = await subject(input, context);
+    expect(actualOutput).toBe(output);
+
+    // IMPLEMENT ME
```
Comment (ritave) [inline test/__tests__/resolvers/FsResolver.spec.ts]: Skip tests

File: test/__tests__/resolvers/GitHubResolver.spec.ts [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+];
+
+describe("GitHubResolver", () => {
+  const subject = GithubResolver();
+
+  it.each(data)("testing %o", async (input, output) => {
+    const actualOutput = await subject(input);
+    expect(actualOutput).toBe(output);
+
+    // IMPLEMENT ME
```
Comment (ritave) [inline test/__tests__/resolvers/GitHubResolver.spec.ts]: Skip test

File: test/__tests__/resolvers/NodeResolver.spec.ts [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+  });
+  afterEach(() => {
+    vol.reset();
+  });
+
+  it.each(data)("testing %o in context %o", async (input, context, output) => {
+    const actualOutput = await subject(input, context);
+    expect(actualOutput).toBe(output);
+
+    // IMPLEMENT ME
```
Comment (ritave) [inline test/__tests__/resolvers/NodeResolver.spec.ts]: Skip test

File: test/__tests__/resolvers/SwarmResolver.spec.ts [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+];
+
+describe("SwarmResolver", () => {
+  const subject = SwarmResolver();
+
+  it.each(data)("testing %o", async (input, output) => {
+    const actualOutput = await subject(input);
+    expect(actualOutput).toBe(output);
+
+    // IMPLEMENT ME
```
Comment (ritave) [inline test/__tests__/resolvers/SwarmResolver.spec.ts]: Skip test

File: test/__tests__/resolvers/IPFSResolver.spec.ts [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+];
+
+describe("IPFSResolver", () => {
+  const subject = IPFSResolver();
+
+  it.each(data)("testing %o", async (input, output) => {
+    const actualOutput = await subject(input);
+    expect(actualOutput).toBe(output);
+
+    // IMPLEMENT ME
```
Comment (ritave) [inline test/__tests__/resolvers/IPFSResolver.spec.ts]: Skip test

File: test/__tests__/resolvers/UriResolver.spec.ts [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+];
+
+describe("UriResolver", () => {
+  const subject = UriResolver();
+
+  it.each(data)("testing %o", async (input, output) => {
+    const actualOutput = await subject(input);
+    expect(actualOutput).toBe(output);
+
+    // IMPLEMENT ME
```
Comment (ritave) [inline test/__tests__/resolvers/UriResolver.spec.ts]: Skip test

File: test/__tests__/solidity/gatherSources.spec.ts [hunk 25]
```diff
@@ -3,7 +3,9 @@ import { vol } from "memfs";
 import { gatherSources, ImportFile, ResolverEngine, SolidityImportResolver } from "../../../src";
 import deepequal = require("deep-equal");
 
-function expectedOutput(filesObj: { [s: string]: string }): ImportFile[] {
+type dictStringString = { [s: string]: string };
```
Comment (ritave) [inline test/__tests__/solidity/gatherSources.spec.ts]: `type Map<K = string, V = string> = { [s: K]: V };`

File: test/__tests__/solidity/gatherSources.spec.ts [hunk 26]
```diff
@@ -23,65 +25,64 @@ function deepSubset<T>(a: T[], b: T[]): boolean {
...
 
-describe("gatherSources function", function() {
-  const resolver: ResolverEngine<ImportFile> = SolidityImportResolver();
-
-  beforeAll(function() {
-    process.chdir(__dirname);
-  });
+// when using mock fs, we are being thrown into the root of the filesystem
+// we need to call it so process.cwd() makes sense
+process.chdir(__dirname);
```
Comment (ritave) [inline test/__tests__/solidity/gatherSources.spec.ts]: This is called in wrong phase - this will chdir ALL the tests, not only this file. Move it into `beforeAll()`

Comment (ritave) [reply test/__tests__/solidity/gatherSources.spec.ts]: If you want data to use cwd, you can set it in beforeAll too

File: test/__tests__/solidity/gatherSources.spec.ts (hunk 25 above)
Comment (cymerrad) [reply test/__tests__/solidity/gatherSources.spec.ts]: I don't understand what this solves. I just took out the type used in the code like three times and gave it some name.

File: src/resolvers/fsresolver.ts (hunk 3 above)
Comment (cymerrad) [reply src/resolvers/fsresolver.ts]: ATM, front-end compiler complains :(

File: test/__tests__/solidity/gatherSources.spec.ts (hunk 26 above)
Comment (cymerrad) [reply test/__tests__/solidity/gatherSources.spec.ts]: 'data' has to be defined, not only declared, globally. It uses '__dirname' now, instead of 'process.cwd()'.

File: src/solidity/ethpmresolver.ts (hunk 13 above)
Comment (cymerrad) [reply src/solidity/ethpmresolver.ts]: I didn't want to call Truffle by name, can I?

File: src/resolvers/noderesolver.ts (hunk 11 above)
Comment (cymerrad) [reply src/resolvers/noderesolver.ts]: I'm not sure we can handle a string that contains shell env variable?

Comment (cymerrad) [reply src/resolvers/noderesolver.ts]: Or what was your intention for expected output?

File: src/resolvers/fsresolver.ts (hunk 3 above)
Comment (ritave) [reply src/resolvers/fsresolver.ts]: It really shouldn't compile node stuff

File: test/__tests__/solidity/gatherSources.spec.ts (hunk 25 above)
Comment (ritave) [reply test/__tests__/solidity/gatherSources.spec.ts]: Names a reason why this is used instead of repeating the declaration itself
```
// this line adds two numbers and sets it to A
let A = 2 + 5;
```
map is easily understood data type, dictstringstring not all that much

Comment (cymerrad) [reply test/__tests__/solidity/gatherSources.spec.ts]: Map in JS implies object with methods like .get(), .set() etc. This is more like Pythonic dictionary, IMHO. I'll go with type name of just 'dictionary'. M'kay?

File: src/resolvers/fsresolver.ts (hunk 3 above)
Comment (cymerrad) [reply src/resolvers/fsresolver.ts]: Can I readdress that in a future merge?

Comment (ritave) [reply src/resolvers/fsresolver.ts]: 👍

**Comment by cymerrad** [comment]
> By the way, why did `ctx` became optional? Two references that are optional already

I've agreed that it will be useful in the future.
